rfcs/0134-nix-store-layer.md

```diff
@@ -0,0 +1,257 @@
+
+Layering between e.g. Flakes and the Nix Language doesn't yet exist in the implementation in the form of a library separation.
+Separating Flakes into its own library would require major new development work, and I don't want to "block" exposing the layering for the first time to users on such work when there is an easier way to try this idea how.
+Flakes also has more development going on due being unstable, so it is nice to not get in its way.
+
+Conversely, the store layer is already quite well separated.
+Its conceptual independence from Nix's other layers is "proven" in the wild by projects like Guix.
+The gap is also widest in terms of the layer of abstraction between, on one hand, nascent "infrastructure projects" like
+
+- content-addressed derivations
+- Secrets in the store
+- Trustless remote building
+- Daemon and Hydra protocol rationalization
+- Windows Support
+- Computed derivations (which are built by the other derivations)
+- IPFS store
+
+and projects aiming at improving user experience, such as
+
+- Flakes
+- TOML Flakes
+- Hard-wired module system
+
+Focusing on the lowest layer first, we get the most "bang for buck" in terms of managing extremely different sorts of work separately.
+
+## Marketplace of Ideas
+
+As Nix grows more popularity, it will be inevitable that different groups want to explore in different directions.
+This is the *pluralism* of a larger community, and we should embrace it as a strength.
+We do that by fostering a *marketplace of ideas*.
+
+There are many possible ways in which to declare packages: the Nix language and Nixpkgs idioms, and Guix, for example, are just two examples of what is possible in principle.
+There are also many possible ways set up build farms.
+Our current model of a central dispatcher, many remote-builder agents, and a client–server where only one side initiates, is also just one point in a much larger design space of possible solutions.
+
+The "derivation language" and store *interface* however, seems to me at least to be a very natural design.
+There are a few tweaks and generalizations we can make, but I struggle to envision what wildly different alternatives one might want instead.
+
+A small, stable interface that allows for design exploration above and below it is known as a [narrow waist](https://www.oilshell.org/cross-ref.html?tag=narrow-waist#narrow-waist).
+It's not every day that a project happens upon a great narrow waist design.
+I believe we've discovered a very good one with Nix, and that should be seen as a *key asset*, even if it is not how we recruit "regular users".
+
+By making a store-only Nix, we put more emphasis on this key interface.
+All functionality the store-only Nix offers factors through this interface.
+The upper half of Nix likewise uses the lower half through this interface.
+The daemon protocol represents this interface for IPC, and allows either half to be swapped for a different implementation.
```

Comment:
So will the daemon protocol need to be standardized as part of this work? it seems to me that it wouldn't make sense to have a defined external executable without a defined way to talk to it (in the long run). The answer could be to link against `libnixstore` and use that to communicate with the (same version of) the store daemon. Is that so?

Reply:
@andir We do have a *defined* deamon protocol, in that you could implement it yourself instead of using `libnixstore` and it is versioned. Yes, it is a not nice, but I don't think improving the situation should be a blocker on this PR.

Rather, I think this PR will help make this sort of thing more noticeable, allowing us to build momentum / community interest about improving the daemon protocol. I think that's great!

Reply:
Yeah, I don't mean this PR should implement it. It is just that the protocol might currently be defined but not documented or carries any guarantees. It has been subtly broken several times and (as you know) it isn't currently being tested across a broad set of nix versions.

I already implemted it (several times in fact) but the protocol isn't fun to reverse engineer from the current code base. My big hope is that with this RFC passing we could move to another RFC that defines a (new) interface that doesn't just allow exchanging implementations but embraces it.

Reply:
> I already implemted it (several times in fact) but the protocol isn't fun to reverse engineer from the current code base.

Might you be interested in documenting these fruits of your protocol-reverse-engineering labors?